File: VB/WinWebSolution.Module/SchedulerActivityListViewControllerBase.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp.Scheduler

Namespace WinWebSolution.Module
	Public MustInherit Class SchedulerActivityListViewControllerBase
		Inherits ViewController(Of ListView)
		Private Shared masterDetailViewEmployeeIdCore As Object
		Private schedulerListEditorCore As SchedulerListEditorBase
		Public Sub New()
			TargetObjectType = GetType(Activity)
		End Sub
		Protected Overrides Overloads Sub OnActivated()
			MyBase.OnActivated()
			If (Not IsAdmininstator(SecuritySystem.CurrentUserId)) Then
				If View.IsRoot Then
					FilterActivities(CriteriaOperator.Parse("Employees[Oid = ?]", SecuritySystem.CurrentUserId))
				End If
			End If
			If (Not View.IsRoot) Then
				Dim propertyCollectionSource As PropertyCollectionSource = TryCast(View.CollectionSource, PropertyCollectionSource)
				If propertyCollectionSource IsNot Nothing Then
					AddHandler propertyCollectionSource.MasterObjectChanged, AddressOf propertyCollectionSource_MasterObjectChanged
				End If
			End If
			schedulerListEditorCore = TryCast((CType(View, ListView)).Editor, SchedulerListEditorBase)
			If schedulerListEditorCore IsNot Nothing Then
				AddHandler schedulerListEditorCore.ResourceDataSourceCreated, AddressOf schedulerListEditorCore_ResourceDataSourceCreated
			End If
		End Sub
		Protected Overrides Overloads Sub OnDeactivating()
			If schedulerListEditorCore IsNot Nothing Then
				RemoveHandler schedulerListEditorCore.ResourceDataSourceCreated, AddressOf schedulerListEditorCore_ResourceDataSourceCreated
			End If
			If (Not View.IsRoot) Then
				Dim propertyCollectionSource As PropertyCollectionSource = TryCast(View.CollectionSource, PropertyCollectionSource)
				If propertyCollectionSource IsNot Nothing Then
					RemoveHandler propertyCollectionSource.MasterObjectChanged, AddressOf propertyCollectionSource_MasterObjectChanged
				End If
			End If
			MyBase.OnDeactivating()
		End Sub
		Private Sub propertyCollectionSource_MasterObjectChanged(ByVal sender As Object, ByVal e As EventArgs)
			Dim masterDetailViewEmployee As Employee = TryCast((CType(sender, PropertyCollectionSource)).MasterObject, Employee)
			If masterDetailViewEmployee IsNot Nothing Then
				masterDetailViewEmployeeIdCore = masterDetailViewEmployee.Oid
				FilterActivities(GetActivitiesFilter(masterDetailViewEmployee))
			End If
		End Sub
		Private Sub schedulerListEditorCore_ResourceDataSourceCreated(ByVal sender As Object, ByVal e As ResourceDataSourceCreatedEventArgs)
			SortResources(e.DataSource)
			FilterResources(e.DataSource, GetResourcesFilter(GetEmployeeById(masterDetailViewEmployeeIdCore)))
		End Sub
		Private Function GetActivitiesFilter(ByVal masterDetailViewEmployee As Employee) As CriteriaOperator
			Dim criteria As CriteriaOperator = Nothing
			If masterDetailViewEmployee IsNot Nothing Then
				If (Not IsAdmininstator(SecuritySystem.CurrentUserId)) Then
					If (View.ObjectSpace.IsNewObject(masterDetailViewEmployee) AndAlso IsAdmininstator(masterDetailViewEmployee)) OrElse ((Not View.ObjectSpace.IsNewObject(masterDetailViewEmployee)) AndAlso IsAdmininstator(masterDetailViewEmployee)) Then
						criteria = CollectionSource.EmptyCollectionCriteria
					End If
				Else
					criteria = CriteriaOperator.Parse("Employees[Oid = ?]", masterDetailViewEmployee.Oid)
				End If
			End If
			Return criteria
		End Function
		Private Function GetResourcesFilter(ByVal masterDetailViewEmployee As Employee) As CriteriaOperator
			Dim criteria As CriteriaOperator = Nothing
			Dim isCurrentUserAdmin As Boolean = IsAdmininstator(SecuritySystem.CurrentUserId)
			Dim isViewRoot As Boolean = View.IsRoot
			Dim isMasterDetailEmployeeNotEmpty As Boolean = masterDetailViewEmployee IsNot Nothing
			If (Not isCurrentUserAdmin) Then
				If (Not isViewRoot) AndAlso isMasterDetailEmployeeNotEmpty Then
					If IsAdmininstator(masterDetailViewEmployee) Then
						criteria = CollectionSource.EmptyCollectionCriteria
					Else
						criteria = CriteriaOperator.Parse("Oid = ?", masterDetailViewEmployee.Oid)
					End If
				ElseIf isViewRoot Then
					criteria = CriteriaOperator.Parse("Oid = ?", SecuritySystem.CurrentUserId)
				End If
			End If
			If isCurrentUserAdmin AndAlso (Not isViewRoot) AndAlso isMasterDetailEmployeeNotEmpty Then
				criteria = CriteriaOperator.Parse("Oid = ?", masterDetailViewEmployee.Oid)
			End If
			Return criteria
		End Function
		Private Function GetEmployeeById(ByVal employeeId As Object) As Employee
			Return Application.CreateObjectSpace().GetObjectByKey(Of Employee)(employeeId)
		End Function
		Private Function IsAdmininstator(ByVal employeeId As Object) As Boolean
			Return IsAdmininstator(GetEmployeeById(employeeId))
		End Function
		Private Shared Function IsAdmininstator(ByVal employee As Employee) As Boolean
			Guard.ArgumentNotNull(employee, "employee")
			Return Convert.ToBoolean(employee.Evaluate(CriteriaOperator.Parse("Groups[Name = ?].Count() > 0", Group.DefaultAdministratorsGroupName)))
		End Function
		Private Sub FilterActivities(ByVal criteria As CriteriaOperator)
			View.CollectionSource.Criteria("FilterActivitiesByEmployee") = criteria
		End Sub
		Protected MustOverride Sub SortResources(ByVal resourcesDataSource As Object)
		Protected MustOverride Sub FilterResources(ByVal resourcesDataSource As Object, ByVal criteria As CriteriaOperator)
	End Class
End Namespace

File: CS/WinWebSolution.Module/SchedulerActivityListViewControllerBase.cs
using System;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp.Scheduler;

namespace WinWebSolution.Module {
    public abstract class SchedulerActivityListViewControllerBase : ViewController<ListView> {
        private static object masterDetailViewEmployeeIdCore;
        private SchedulerListEditorBase schedulerListEditorCore;
        public SchedulerActivityListViewControllerBase() {
            TargetObjectType = typeof(Activity);
        }
        protected override void OnActivated() {
            base.OnActivated();
            if (!IsAdmininstator(SecuritySystem.CurrentUserId)) {
                if (View.IsRoot) {
                    FilterActivities(CriteriaOperator.Parse("Employees[Oid = ?]", SecuritySystem.CurrentUserId));
                }
            }
            if (!View.IsRoot) {
                PropertyCollectionSource propertyCollectionSource = View.CollectionSource as PropertyCollectionSource;
                if (propertyCollectionSource != null) {
                    propertyCollectionSource.MasterObjectChanged += propertyCollectionSource_MasterObjectChanged;
                }
            }
            schedulerListEditorCore = ((ListView)View).Editor as SchedulerListEditorBase;
            if (schedulerListEditorCore != null) {
                schedulerListEditorCore.ResourceDataSourceCreated += schedulerListEditorCore_ResourceDataSourceCreated;
            }
        }
        protected override void OnDeactivating() {
            if (schedulerListEditorCore != null) {
                schedulerListEditorCore.ResourceDataSourceCreated -= schedulerListEditorCore_ResourceDataSourceCreated;
            }
            if (!View.IsRoot) {
                PropertyCollectionSource propertyCollectionSource = View.CollectionSource as PropertyCollectionSource;
                if (propertyCollectionSource != null) {
                    propertyCollectionSource.MasterObjectChanged -= propertyCollectionSource_MasterObjectChanged;
                }
            }
            base.OnDeactivating();
        }
        private void propertyCollectionSource_MasterObjectChanged(object sender, EventArgs e) {
            Employee masterDetailViewEmployee = ((PropertyCollectionSource)sender).MasterObject as Employee;
            if (masterDetailViewEmployee != null) {
                masterDetailViewEmployeeIdCore = masterDetailViewEmployee.Oid;
                FilterActivities(GetActivitiesFilter(masterDetailViewEmployee));
            }
        }
        private void schedulerListEditorCore_ResourceDataSourceCreated(object sender, ResourceDataSourceCreatedEventArgs e) {
            SortResources(e.DataSource);
            FilterResources(e.DataSource, GetResourcesFilter(GetEmployeeById(masterDetailViewEmployeeIdCore)));
        }
        private CriteriaOperator GetActivitiesFilter(Employee masterDetailViewEmployee) {
            CriteriaOperator criteria = null;
            if (masterDetailViewEmployee != null) {
                if (!IsAdmininstator(SecuritySystem.CurrentUserId)) {
                    if ((View.ObjectSpace.IsNewObject(masterDetailViewEmployee) && IsAdmininstator(masterDetailViewEmployee))
                            || (!View.ObjectSpace.IsNewObject(masterDetailViewEmployee) && IsAdmininstator(masterDetailViewEmployee))) {
                        criteria = CollectionSource.EmptyCollectionCriteria;
                    }
                }
                else {
                    criteria = CriteriaOperator.Parse("Employees[Oid = ?]", masterDetailViewEmployee.Oid);
                }
            }
            return criteria;
        }
        private CriteriaOperator GetResourcesFilter(Employee masterDetailViewEmployee) {
            CriteriaOperator criteria = null;
            bool isCurrentUserAdmin = IsAdmininstator(SecuritySystem.CurrentUserId);
            bool isViewRoot = View.IsRoot;
            bool isMasterDetailEmployeeNotEmpty = masterDetailViewEmployee != null;
            if (!isCurrentUserAdmin) {
                if (!isViewRoot && isMasterDetailEmployeeNotEmpty) {
                    if (IsAdmininstator(masterDetailViewEmployee)) {
                        criteria = CollectionSource.EmptyCollectionCriteria;
                    }
                    else {
                        criteria = CriteriaOperator.Parse("Oid = ?", masterDetailViewEmployee.Oid);
                    }
                }
                else if (isViewRoot) {
                    criteria = CriteriaOperator.Parse("Oid = ?", SecuritySystem.CurrentUserId);
                }
            }
            if (isCurrentUserAdmin && !isViewRoot && isMasterDetailEmployeeNotEmpty) {
                criteria = CriteriaOperator.Parse("Oid = ?", masterDetailViewEmployee.Oid);
            }
            return criteria;
        }
        private Employee GetEmployeeById(object employeeId) {
            return Application.CreateObjectSpace().GetObjectByKey<Employee>(employeeId);
        }
        private bool IsAdmininstator(object employeeId) {
            return IsAdmininstator(GetEmployeeById(employeeId));
        }
        private static bool IsAdmininstator(Employee employee) {
            Guard.ArgumentNotNull(employee, "employee");
            return Convert.ToBoolean(employee.Evaluate(CriteriaOperator.Parse("Groups[Name = ?].Count() > 0", Group.DefaultAdministratorsGroupName)));
        }
        private void FilterActivities(CriteriaOperator criteria) {
            View.CollectionSource.Criteria["FilterActivitiesByEmployee"] = criteria;
        }
        protected abstract void SortResources(object resourcesDataSource);
        protected abstract void FilterResources(object resourcesDataSource, CriteriaOperator criteria);
    }
}